module6/lessons/Week5/Recursion.md

```diff
@@ -0,0 +1,178 @@
+
+Recursion is typically used when an operation needs to be completed repeatedly, with different parameters.  It is most effective for solving problems around sorting, traversing nodes, and fractal math.  
+
+Some things to consider when determining if you should use a recursive function:  
+✅there is no need to keep track of state; there are no side effects _outside_ of the recursive function.  
+✅optimizes traversal of non-linear datatypes.  
+❌Execution contexts continue to get build on the call stack; with bigger datasets, you could run into `maximum call stack size being exceeded`  
+
+## Functions and the Call Stack
+
+**Stack Overflow** As developers, especially junior developers, when you see these words, you are probably imagining the website stackoverflow.com.  The inspiration for this website is a real concept - the **call stack**.
+
+A **callstack** is at the heart of how programs are executed - it keeps track of the subroutines (methods, functions) that have started, but haven't yet finished.
+
+Let's discuss what happens on the stack when we execute methods!
+
+<section class='instructor-notes' markdown='1'>
+
+Use the slides below to walkthrough what is happening on the stack.  Not _everything_ that is happening on the stack is specifically called out - I tried to include only the most necessary information.
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### Diagramming the Call Stack
+
+<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vRFJ2pNfsOap3jgooinxMqFWbcKASjPv8b0TY_lHCh53vD3ts0DE_fGqQuJclIaWkctq-qfLtZjJ0r0/embed?start=false&loop=false&delayms=60000" frameborder="0" width="960" height="569" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true"></iframe>
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### When Disaster Strikes
+
+<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vTM6N3zr7xzZPws6wzoIvNLduMQwVMg-AItpcIRrNfqoAVoY8tA3ibd0bHMdCebh0bFAFQlhfEroY0B/embed?start=false&loop=false&delayms=60000" frameborder="0" width="960" height="569" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true"></iframe>
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### Recursive Functions
+
+<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vQTd07os2suundP21vi48fsSPO--NOzFnLHgMxgWODyvPEAVwCchApJLR10vQq3kK99OEmXmZGto_qL/embed?start=false&loop=false&delayms=60000" frameborder="0" width="960" height="569" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true"></iframe>
+
+</section>
+
+## Exercises
+
+The best way to start understanding recursion is to just try doing it!  Feel free to work through these problems in any language.
+
+### Exercise 1
+
+Reverse a string.
+
+```js
+// create a function which takes a string of characters and
+// recursively calls itself to reverse the string
+
+// e.g.
+
+string reversedString = reverse('Ariel');
```

Comment:
I might add an instructor note here to talk as a class through the base case and the recursive case (how does the recursive case get closer to the base case). Or maybe do a breakout room to talk with your partner and then write the pseudo code as a class. I think starting from pseudocode for the first one will be helpful!